### ISSUE TITLE
Feature Request: Add Option to use Improved Versions of CLIP such as MetaCLIP, SigLip, CLIPA

### ISSUE THREAD

**Issue by shelbywhite** [weaviate/multi2vec-clip-inference#32]
Feature Request:
Would love to be able to drop in any of the [MetaClip](https://github.com/facebookresearch/MetaCLIP) models. [These models](https://huggingface.co/models?other=metaclip) seem to improve on accuracy over the base CLIP model. Any of these models would be great, but the option to choose which of the many MetaClip models would be ideal. As for which model is best, it would really be a comparison of inference vs accuracy in comparison to the base version of clip. 

In my system I am bound to CPUs and inference is the issue. My Weaviate set up is manually installed and does not use the traditional Docker set up. All in all looking for a way to improve inference so that a higher performing model could be used. ImageBind is great, but inference is slow on CPU. 

Another option here would be to see if ONNX models help with inference. This isn't something I have done yet.

To see which model _might_ make the most sense, I have created a _very rudimentary_ script to test Model & Processor Load Times as well as Inference time averaged over 5 images. That script is included below. After running this many times, while the times change slightly, they generally appear in the same order. What we see is that the highest quality models seem to be the slowest inference which is to be expected. 

<img width="812" alt="Screenshot 2024-03-08 at 1 38 22 PM" src="https://github.com/weaviate/multi2vec-clip-inference/assets/252104/723fd7c2-0cc0-490f-aeb9-b1e5ee7498a6">

<img width="849" alt="Screenshot 2024-03-08 at 1 29 14 PM" src="https://github.com/weaviate/multi2vec-clip-inference/assets/252104/ed4f1f88-f694-42f4-a5e5-1c23ce8fb32d">

### Script to test inference timing:
```
import torch
import time
import glob
import csv
import gc
from PIL import Image
from transformers import AutoProcessor, AutoModel

models = [
	'openai/clip-vit-base-patch32',
	'facebook/metaclip-b32-400m',
	'facebook/metaclip-b32-fullcc2.5b',
	'facebook/metaclip-h14-fullcc2.5b',
	'facebook/metaclip-b16-fullcc2.5b',
	'facebook/metaclip-b16-400m',
	'facebook/metaclip-l14-fullcc2.5b',
	'facebook/metaclip-l14-400m',
]

with open('model_analysis.csv', mode='w', newline='') as file:
	writer = csv.writer(file)
	writer.writerow(['Model Name', 'Average Processor Loading Time (s)', 'Average Model Loading Time (s)', 'Average Inference Time (s)'])

	for model_name in models:
		processor_loading_times = []
		model_loading_times = []

		# Measure processor and model loading times
		for _ in range(5):
			start_time_processor = time.time()
			processor = AutoProcessor.from_pretrained(model_name)
			end_time_processor = time.time()
			processor_loading_times.append(end_time_processor - start_time_processor)

			start_time_model = time.time()
			model = AutoModel.from_pretrained(model_name)
			end_time_model = time.time()
			model_loading_times.append(end_time_model - start_time_model)

		# Glob match image files
		image_files = glob.glob('images/*')

		# List to store inference times
		inference_times = []

		# Process the first 5 images
		for image_path in image_files[:5]:
			image = Image.open(image_path)
			inputs = processor(text=[''], images=image, return_tensors="pt", padding=True)

			# Start timing
			start_time = time.time()

			# Inference
			with torch.no_grad():
				outputs = model(**inputs)
				image_embeddings = outputs.image_embeds 
				# text_embeddings = outputs.text_embeds 

			# End timing
			end_time = time.time()

			# Calculate and store the inference time for this image
			inference_time = end_time - start_time
			inference_times.append(inference_time)

		# Calculate the average inference time
		average_inference_time = sum(inference_times) / len(inference_times)
		average_processor_loading_time = sum(processor_loading_times) / len(processor_loading_times)
		average_model_loading_time = sum(model_loading_times) / len(model_loading_times)

		# Write data to CSV
		writer.writerow([model_name, average_processor_loading_time, average_model_loading_time, average_inference_time])

		print(f"Model: {model_name}")
		print(f"Average Processor Loading Time: {average_processor_loading_time:.4f} seconds")
		print(f"Average Model Loading Time: {average_model_loading_time:.4f} seconds")
		print(f"Average Inference Time: {average_inference_time:.4f} seconds")
		print('---------------------------------')

		# After processing each model, attempt to free up memory
		del model
		del processor
		if torch.cuda.is_available():
			torch.cuda.empty_cache()

		gc.collect()
```

More information about how MetaCLIP came to be—Demystifying CLIP Data:
https://arxiv.org/pdf/2309.16671.pdf

![1698428214641](https://github.com/weaviate/multi2vec-clip-inference/assets/252104/e79f1a0d-d092-49b8-8526-6f281bafb66f)
